File: design-patterns/golang/state/has_money_state.go
package state

import "fmt"

// HasMoneyState represents the state after money is inserted
type HasMoneyState struct{}

func (s *HasMoneyState) SelectItem(context *VendingMachine, item string) {
	fmt.Println("Transaction in progress. Please wait.")
}

func (s *HasMoneyState) InsertCoin(context *VendingMachine, amount float64) {
	fmt.Println("Already received money. Dispensing item...")
}

func (s *HasMoneyState) DispenseItem(context *VendingMachine) {
	item := context.GetSelectedItem()
	amt := context.GetInsertedAmount()
	fmt.Printf("Dispensing '%s' for $%.2f. Thank you!\n", item, amt)
	context.Reset()
}

File: solutions/typescript/src/VendingMachine/VendingMachineState/HasMoneyState.ts
import VendingMachine from "../VendingMachine";
import VendingMachineState from "./VendingMachineState";

export default class HasMoneyState implements VendingMachineState {
  private machine: VendingMachine;

  constructor(machine: VendingMachine) {
    this.machine = machine;
  }

  insertCoin() {
    console.log("Already received full amount.");
  }

  selectItem() {
    console.log("Item already selected.");
  }

  dispense() {
    this.machine.setState(this.machine.getDispensingState());
    this.machine.dispenseItem();
  }

  refund() {
    this.machine.refundBalance();
    this.machine.resetToIdleState();
    this.machine.setState(this.machine.getIdleState());
  }
}
